File: examples/agent-framework-webhooks/python/agent_webhooks/adapters/adk.py
"""Google Agent Development Kit reply adapter."""

import hashlib
import os
from collections.abc import AsyncIterable, Callable
from email.utils import parseaddr
from typing import Any

from e2a import AsyncInboundEmail

from agent_webhooks.prompt import REPLY_INSTRUCTIONS, email_prompt

APP_NAME = "e2a_email_assistant"
ADKRun = Callable[[AsyncInboundEmail, str, str], AsyncIterable[Any]]


def _user_id(email: AsyncInboundEmail) -> str:
    sender = parseaddr(email.from_ or "")[1].strip().casefold() or "missing-sender"
    inbox = email.inbox.strip().casefold() or "missing-inbox"
    namespace = f"{inbox}\0{sender}"
    digest = hashlib.sha256(namespace.encode("utf-8")).hexdigest()[:20]
    return f"sender-{digest}"


class ADKReplyAgent:
    def __init__(self, run: ADKRun) -> None:
        self._run = run

    @classmethod
    def from_env(cls) -> "ADKReplyAgent":
        from google.adk.agents import LlmAgent
        from google.adk.runners import Runner
        from google.adk.sessions import InMemorySessionService
        from google.genai import types

        agent = LlmAgent(
            name="email_assistant",
            model=os.getenv("ADK_MODEL", "gemini-flash-latest"),
            instruction=REPLY_INSTRUCTIONS,
        )
        sessions = InMemorySessionService()
        runner = Runner(agent=agent, app_name=APP_NAME, session_service=sessions)

        async def run(
            email: AsyncInboundEmail, prompt: str, conversation_id: str
        ) -> AsyncIterable[Any]:
            user_id = _user_id(email)
            session_id = conversation_id
            session = await sessions.get_session(
                app_name=APP_NAME,
                user_id=user_id,
                session_id=session_id,
            )
            if session is None:
                await sessions.create_session(
                    app_name=APP_NAME,
                    user_id=user_id,
                    session_id=session_id,
                )
            content = types.Content(
                role="user",
                parts=[types.Part(text=prompt)],
            )
            async for event in runner.run_async(
                user_id=user_id,
                session_id=session_id,
                new_message=content,
            ):
                yield event

        return cls(run)

    async def reply(
        self, email: AsyncInboundEmail, conversation_id: str
    ) -> str:
        final_text = ""
        async for event in self._run(email, email_prompt(email), conversation_id):
            if not event.is_final_response():
                continue
            content = getattr(event, "content", None)
            parts = getattr(content, "parts", []) if content is not None else []
            final_text = "\n".join(
                part.text
                for part in parts
                if isinstance(getattr(part, "text", None), str)
            )
        return final_text

File: examples/agent-framework-webhooks/typescript/src/adapters/adk.ts
import { createHash } from "node:crypto";

import type { InboundEmail } from "@e2a/sdk/v1";
import { InMemoryRunner, LlmAgent, isFinalResponse, type Event } from "@google/adk";

import type { ReplyAgent } from "../contracts.js";
import { REPLY_INSTRUCTIONS, emailPrompt } from "../prompt.js";

export const ADK_APP_NAME = "e2a_email_assistant";

export interface ADKRunInput {
  prompt: string;
  userId: string;
  sessionId: string;
}

export type ADKRun = (input: ADKRunInput) => AsyncIterable<Event>;

function canonicalMailbox(sender: string | null): string {
  const value = (sender ?? "").trim();
  const angleAddress = value.match(/<\s*([^<>\s]+@[^<>\s]+)\s*>/u)?.[1];
  return (angleAddress ?? value).trim().toLowerCase() || "missing-sender";
}

/** Derive an opaque identity while preventing session sharing across e2a inboxes. */
export function senderUserId(email: Pick<InboundEmail, "from" | "inbox">): string {
  const inbox = email.inbox.trim().toLocaleLowerCase("en-US") || "missing-inbox";
  const namespace = `${inbox}\0${canonicalMailbox(email.from)}`;
  const digest = createHash("sha256").update(namespace, "utf8").digest("hex").slice(0, 20);
  return `sender-${digest}`;
}

function eventText(event: Event): string {
  return event.content?.parts?.flatMap((part) =>
    typeof part.text === "string" ? [part.text] : [],
  ).join("\n") ?? "";
}

export class ADKReplyAgent implements ReplyAgent {
  constructor(private readonly run: ADKRun) {}

  async reply(email: InboundEmail, conversationId: string): Promise<string> {
    let reply = "";
    const input = {
      prompt: emailPrompt(email),
      userId: senderUserId(email),
      sessionId: conversationId,
    };
    for await (const event of this.run(input)) {
      if (isFinalResponse(event)) reply = eventText(event);
    }
    return reply;
  }
}

/** Build the production adapter from the official `@google/adk` exports. */
export function createADKReplyAgent(
  env: Record<string, string | undefined> = process.env,
): ADKReplyAgent {
  const agent = new LlmAgent({
    name: "email_assistant",
    model: env.ADK_MODEL ?? "gemini-flash-latest",
    instruction: REPLY_INSTRUCTIONS,
  });
  const runner = new InMemoryRunner({ agent, appName: ADK_APP_NAME });

  return new ADKReplyAgent(async function* ({ prompt, userId, sessionId }) {
    const sessionContext = {
      appName: ADK_APP_NAME,
      userId,
      sessionId,
    };
    const session = await runner.sessionService.getSession(sessionContext);
    if (session === undefined) await runner.sessionService.createSession(sessionContext);
    yield* runner.runAsync({
      userId,
      sessionId,
      newMessage: { role: "user", parts: [{ text: prompt }] },
    });
  });
}
